File: AI_Challenges/AI_Challenges_TN/challenge-10/src/scorer.ts
import type { RuleFlag, ScoredClaim } from './types.js'
import { MAX_POSSIBLE } from './constants.js'

export function scoreClaim(claim_id: string, flags: RuleFlag[]): ScoredClaim {
  const triggered = flags.filter(f => f.triggered)
  const weightedSum = triggered.reduce((s, f) => s + f.severity, 0)
  const score = Math.min(100, Math.round((weightedSum / MAX_POSSIBLE) * 100))
  return { claim_id, score, flags: triggered }
}

export function scoreAll(claimFlags: Map<string, RuleFlag[]>): ScoredClaim[] {
  const scored: ScoredClaim[] = []
  for (const [claim_id, flags] of claimFlags) {
    scored.push(scoreClaim(claim_id, flags))
  }
  return scored.sort((a, b) => b.score - a.score)
}

File: AI_Challenges/AI_Challenges_TN/challenge-10/scorer.py
"""Composite risk scorer. Severity weights justified in README."""

# Weights reflect signal strength and fraud certainty:
# - duplicate/phantom = definitive fraud patterns → 5
# - upcoding/weekend/mismatch = strong statistical/clinical evidence → 4
# - rapid_resubmission/unbundling = indicative but legitimately explainable → 3
# - amount_clustering = weak alone (could be coincidence) → 2
SEVERITY_WEIGHTS = {
    'duplicate_claim':     5,
    'phantom_billing':     5,
    'upcoding':            4,
    'weekend_anomaly':     4,
    'dx_proc_mismatch':    4,
    'rapid_resubmission':  3,
    'unbundling':          3,
    'amount_clustering':   2,
}

MAX_POSSIBLE = sum(SEVERITY_WEIGHTS.values())  # 30 — used for normalization


def score_claim(flags: list) -> int:
    """0–100 composite score, capped at 100."""
    if not flags:
        return 0
    total = sum(SEVERITY_WEIGHTS.get(f['rule'], f.get('severity', 1)) for f in flags)
    return min(100, round(total / MAX_POSSIBLE * 100))
